### PR TITLE
Feature/dialogues offload implementation

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

Offloading feature for PersistsDialoguesStorage

## Types of changes
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/protocols/dialogue/base.py [hunk 1]
```diff
@@ -1210,30 +1227,36 @@ def _get_collection_instance(self, col_name: str) -> Optional[SyncCollection]:
...
-    ) -> Tuple[Optional[SyncCollection], Optional[SyncCollection]]:
-        """Get active state dialogue collection and terminal collection."""
+    @cached_property
+    def _terminal_dialogues_collection(self) -> Optional[SyncCollection]:
         col_name = self._get_collection_name()
         if not col_name:
-            return (None, None)  # pragma: nocover
-        termina_col_name = f"{col_name}_terminal"
+            return None
+        col_name += "_terminal"
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1235]: define constant elsewhere and reuse here

File: aea/protocols/dialogue/base.py [hunk 2]
```diff
@@ -1309,6 +1344,121 @@ def teardown(self) -> None:
...
+            collection = self._active_dialogues_collection
+
+        super().remove(dialogue_label)
+
+        if collection:
+            collection.remove(str(dialogue_label))
+
+
+class PersistDialoguesStorageWithOffloading(PersistDialoguesStorage):
+    """Dialogue Storage with dialogues offlosding."""
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1361]: typo

File: aea/protocols/dialogue/base.py [hunk 3]
```diff
@@ -1309,6 +1344,121 @@ def teardown(self) -> None:
...
+                    super().get_dialogues_with_counterparty(counterparty)
+                    + self._get_dialogues_by_address_from_collection(
+                        counterparty, self._active_dialogues_collection
+                    )
+                    + self._get_dialogues_by_address_from_collection(
+                        counterparty, self._terminal_dialogues_collection
+                    )
+                )
+            }.values()
+        )
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1452]: refactor, hard to read

File: aea/protocols/dialogue/base.py [hunk 4]
```diff
@@ -1309,6 +1344,121 @@ def teardown(self) -> None:
...
+
+    @property
+    def dialogues_in_terminal_state(self) -> List["Dialogue"]:
+        """Get all dialogues in terminal state."""
+        dialogues = super().dialogues_in_terminal_state + list(
+            self._load_dialogues(self._terminal_dialogues_collection)
+        )
+        return list({i.dialogue_label: i for i in dialogues}.values())
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1460]: don't use single char vars; just call it `dialogue`

Comment (DavidMinarsch) [reply aea/protocols/dialogue/base.py:1460]: Also, what's the point of this line?

File: aea/protocols/dialogue/base.py [hunk 5]
```diff
@@ -1309,6 +1344,121 @@ def teardown(self) -> None:
...
+        dialogue = super().get(dialogue_label)
+        if dialogue:
+            return dialogue
+
+        dialogue = self._get_dialogue_from_collection(
+            dialogue_label, self._terminal_dialogues_collection
+        )
+        if dialogue:
+            # get dialogue from terminal state collection and cache it
+            self._add_terminal_state_dialogue(dialogue)
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1388]: why?

File: aea/protocols/dialogue/base.py [hunk 6]
```diff
@@ -1270,19 +1298,26 @@ def _dump_dialogues(
             collection.put(str(dialogue.dialogue_label), dialogue.json())
 
     def _load(self) -> None:
-        """Load dialogues and incomplete dialogues from the generic storage."""
-        active_collection, terminal_collection = self._get_collections()
-        if not active_collection or not terminal_collection:
+        """Dump dialogues and incomplete dialogues from the generic storage."""
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1301]: typo

File: aea/protocols/dialogue/base.py [hunk 7]
```diff
@@ -1253,13 +1276,18 @@ def _load_incomplete_dialogues_labels(self, collection: SyncCollection) -> None:
 
     def _load_dialogues(self, collection: SyncCollection) -> Iterable[Dialogue]:
         """Load dialogues from collection."""
+        if not collection:  # pragma: nocover
+            return
```
Comment (DavidMinarsch) [inline aea/protocols/dialogue/base.py:1280]: None return type not accounted for in typing

File: tests/test_protocols/test_dialogue/test_base.py [hunk 8]
```diff
@@ -1810,22 +1812,171 @@ def test_dump_restore(self):
...
+
+        assert len(dialogues_by_addr) == len(
+            dialogues_storage_restored.get_dialogues_with_counterparty(
+                dialogue.dialogue_label.dialogue_opponent_addr
+            )
+        )
+        # check get and cache
+        assert not dialogues_storage_restored._terminal_state_dialogues_labels
+        assert dialogues_storage_restored.get(dialogue_label)
+        assert dialogues_storage_restored._terminal_state_dialogues_labels
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_dialogue/test_base.py:1965]: I'd prefer the asserts to be more explicit!

File: aea/helpers/base.py [hunk 9]
```diff
@@ -499,3 +498,61 @@ def reachable_nodes(
         queue.extendleft(successors)
         visited.add(current)
     return result
+
+
+_NOT_FOUND = object()
+
+
+# copied from python3.8 functools
+class cached_property:  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/helpers/base.py:507]: Not available in < 3.8?

Comment (solarw) [reply aea/helpers/base.py:507]: yes

File: aea/protocols/dialogue/base.py (hunk 5 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1388]: just to avoid direct access to internal valirables

File: aea/protocols/dialogue/base.py (hunk 2 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1361]: done

File: aea/protocols/dialogue/base.py (hunk 1 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1235]: done

File: aea/protocols/dialogue/base.py (hunk 6 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1301]: but it looks ok

File: aea/protocols/dialogue/base.py (hunk 7 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py:1280]: in this case it's the end of the generator, so it's ok

File: tests/test_protocols/test_dialogue/test_base.py (hunk 8 above)
Comment (solarw) [reply tests/test_protocols/test_dialogue/test_base.py:1965]: slightly updated

File: aea/protocols/dialogue/base.py (hunk 3 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py]: done

File: aea/protocols/dialogue/base.py (hunk 4 above)
Comment (solarw) [reply aea/protocols/dialogue/base.py]: discussed

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**